crates/sui-core/src/epoch/randomness.rs

```diff
@@ -0,0 +1,516 @@
+            }
+        };
+        let (nodes, t) = nodes.reduce(
+            committee
+                .validity_threshold()
+                .try_into()
+                .expect("validity threshold should fit in u16"),
+            protocol_config.random_beacon_reduction_allowed_delta(),
+        );
+        let total_weight = nodes.total_weight();
+        let num_nodes = nodes.num_nodes();
+        let prefix_str = format!(
+            "dkg {} {}",
+            Hex::encode(epoch_store.get_chain_identifier().as_bytes()),
+            committee.epoch()
+        );
+        let randomness_private_key = bls12381::Scalar::from_byte_array(
+            authority_key_pair
+                .copy()
+                .private()
+                .as_bytes()
+                .try_into()
+                .expect("key length should match"),
+        )
+        .expect("should work to convert BLS key to Scalar");
+        let party = match dkg::Party::<PkG, EncG>::new(
+            fastcrypto_tbls::ecies::PrivateKey::<bls12381::G2Element>::from(randomness_private_key),
+            nodes,
+            t.into(),
+            fastcrypto_tbls::random_oracle::RandomOracle::new(prefix_str.as_str()),
+            &mut rand::thread_rng(),
+        ) {
+            Ok(party) => party,
+            Err(err) => {
+                error!("random beacon: error while initializing Party: {err:?}");
+                return None;
+            }
+        };
+        info!(
+            "random beacon: state initialized with authority={name}, total_weight={total_weight}, t={t}, num_nodes={num_nodes}, oracle initial_prefix={prefix_str:?}",
+        );
+
+        // Load existing data from store.
+        let manager = Self {
+            epoch_store: epoch_store_weak,
+            consensus_adapter,
+            party,
+            processed_messages: Mutex::new(BTreeMap::new()),
+            used_messages: OnceCell::new(),
+            confirmations: Mutex::new(BTreeMap::new()),
+            dkg_output: OnceCell::new(),
+        };
+        let dkg_output = tables
+            .dkg_output
+            .get(&SINGLETON_KEY)
+            .expect("typed_store should not fail");
+        if let Some(dkg_output) = dkg_output {
+            info!(
+                "random beacon: loaded existing DKG output for epoch {}",
+                committee.epoch()
+            );
+            epoch_store
+                .metrics
+                .epoch_random_beacon_dkg_num_shares
+                .set(dkg_output.shares.as_ref().map_or(0, |shares| shares.len()) as i64);
+            manager
+                .dkg_output
+                .set(dkg_output)
+                .expect("setting new OnceCell should succeed");
+        } else {
+            info!(
+                "random beacon: no existing DKG output found for epoch {}",
+                committee.epoch()
+            );
+            // Load intermediate data.
+            manager.processed_messages.lock().unwrap().extend(
```

Comment:
what happens if that list is extended after this line is executed?

Reply:
Not sure I follow the question? This section of the code is for recovery of state post-restart

It might be helpful for understanding to note that in spite of the Mutexes this code is only executed single-threaded, the Mutexes are there because the functions that call into `RandomnessManager` all use `&self` and not `&mut self` and it seemed like a big lift to propagate mutable references all the way up through `AuthorityPerEpochStore` - I could try it but I'm not sure how feasible that even is? Curious if @mystenmark has any opinions there.

Reply:
I would consider putting the mutable state into a `RandomnessInner` struct, which is held in a mutex, and whose methods are all `&mut self` - this way you have convenient interior mutability (no viral `&mut self`) but it is clearer that the lock protects all the state at once.

Reply:
I switched to this pattern, thanks for the suggestion